src/lib/core/datetime.h

```diff
@@ -70,6 +70,33 @@ struct datetime {
 	int16_t tzindex;
 };
 
+/**
+ * To be able to perform arithmetics on time intervals and receive
+ * deterministic results, we have to keep months and years separately
+ * from seconds.
+ * Weeks, days, hours and minutes, all could be _precisely_ converted
+ * to seconds, but it's not the case for months (which might be 28, 29,
+ * 30, or 31 days long), or years (which could be leap year or not).
+ * Approach used here - to add/subtract months or years intervals only
+ * at the moment when we have particular date we operate on.
+ * Determinism of results is achieved due to the order we apply
+ * operations (from larger to smaller quantities):
+ * - years, then months, then weeks, days, hours, minutes,
+ *   seconds, and nanoseconds.
+*/
+struct datetime_interval {
+	/** duration in seconds  */
+	double sec;
+	/** fraction part of duration in seconds */
+	int nsec;
+	/** number of months, if specified */
+	int month;
+	/** number of years, if specified */
+	int year;
+	/** day in month adjustment mode, @sa dt_adjust */
+	int /*dt_adjust_t*/ adjust;
```

Comment:
I could not defined enum here - it's defined in the `c-dt` afterwards. Originally I tried to avoid inserting extra dependency in this header, but you've highlighted the correct case - it makes no much sense in `datetime.h` to avoid `c-dt` dependency, thus I've simplified here a bit and got rid of this trick:

```diff
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -9,7 +9,6 @@
 #include <string.h>
 #include <time.h>
 
-#include "c-dt/dt.h"
 #include "datetime.h"
 #include "trivia/util.h"
 #include "tzcode/tzcode.h"
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index fa6313c6c..7437550e3 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -8,6 +8,7 @@
 #include <limits.h>
 #include <stdint.h>
 #include <sys/types.h>
+#include "c-dt/dt.h"
 
 #if defined(__cplusplus)
 extern "C" {
@@ -101,8 +102,8 @@ struct datetime_interval {
 	int month;
 	/** Number of years, if specified. */
 	int year;
-	/** Adjustment mode for day in month operations, @sa dt_adjust */
-	int /*dt_adjust_t*/ adjust;
+	/** Adjustment mode for day in month operations, @sa dt_adjust_t */
+	dt_adjust_t adjust;
 };
 
 /*
diff --git a/src/lua/utils.c b/src/lua/utils.c
index b47c586f3..33fe074a8 100644
--- a/src/lua/utils.c
+++ b/src/lua/utils.c
@@ -755,7 +755,7 @@ tarantool_lua_utils_init(struct lua_State *L)
 			  "int nsec;"
 			  "int month;"
 			  "int year;"
-			  "int adjust;"
+			  "dt_adjust_t adjust;"
 			  "};");
 	assert(rc == 0);
 	(void) rc;

```